inference-engine/src/mkldnn_plugin/nodes/ctc_loss.cpp

```diff
@@ -0,0 +1,338 @@
+
+        I2 blankIndex = classesNum - 1;
+        if (inputs.size() > 4) {
+            blankIndex = inputs[4]->cbuffer().as<const I2*>()[0];
+        }
+
+        std::vector<I2> targetD(maxTime);
+
+        const size_t TC = maxTime * classesNum;
+
+        for (size_t b = 0; b < batchNum; b++) {
+            const I1 actualLogitLen = logitsLength[b];
+            const I1 actualTargetLen = labelsLength[b];
+            if (actualLogitLen > maxTime || actualTargetLen > maxTime || actualTargetLen > actualLogitLen) {
+                std::string errorMsg = _logPrefix + ". Logit or label length cannot greater than max sequence length. "
+                    + "Also a label length cannot be greater than a logit length.\nMaxSeqLen: "
+                    + std::to_string(maxTime) + "; Logit len: " + std::to_string(actualLogitLen)
+                    + "; Label len: " + std::to_string(actualTargetLen);
+                errorMsg.copy(resp->msg, sizeof(resp->msg) - 1);
+                return GENERAL_ERROR;
+            }
+
+            const I2* target = &labels[b * maxTime];
+            // Decoding target: merge repeated characters if preprocess_collapse_repeated == True,
+            // find unique elemnts if unique == True
+            size_t decodedTargetLen = 0lu;
+            if (_unique) {
+                std::unordered_set<I2> uniqVals;
+                for (size_t t = 0lu; t < actualTargetLen; t++) {
+                    if (uniqVals.find(target[t]) != uniqVals.end()) {
+                        continue;
+                    }
+                    uniqVals.insert(target[t]);
+                    targetD[decodedTargetLen++] = target[t];
+                }
+            } else if (_preprocessCollapseRepeated) {
+                I2 prevValue = target[0];
+                targetD[decodedTargetLen++] = target[0];
+                for (size_t t = 1lu; t < actualTargetLen; t++) {
+                    if (target[t] == prevValue) {
+                        continue;
+                    }
+                    targetD[decodedTargetLen++] = target[t];
+                    prevValue = target[t];
+                }
+            } else {
+                std::copy(target, target + actualTargetLen, targetD.data());
+                decodedTargetLen = actualTargetLen;
+            }
+
+            const size_t BTC = b * TC;
+
+            std::vector<std::unordered_map<size_t, float>> probabilities(actualLogitLen);
+            float prob = 0.f, kExp = 0.f;
+            for (size_t t = 0; t < actualLogitLen; t++) {
+                kExp = 0.f;
+                const size_t btcT = BTC + classesNum * t;
+                for (size_t c = 0; c < classesNum; c++) {
+                    kExp += std::exp(logits[btcT + c]);
+                }
+                for (size_t s = 0; s < decodedTargetLen; s++) {
+                    prob = std::log(std::exp(logits[btcT + targetD[s]]) / kExp);
+                    probabilities[t].insert({targetD[s], prob});
```

Comment:
Here I recommend to use log softmax formula that is more stable numerically

Reply:
Fixed